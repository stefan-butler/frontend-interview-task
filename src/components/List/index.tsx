import React, { useEffect, useState } from 'react'
import { Allowance } from '@/types/allowances'
import Card from '../Card';

const List = () => {
  const [allowances, setAllowances] = useState<Allowance[]>([])

  useEffect(() => {
    const fetchData = async () => {
      const response = await fetch('/api/allowances')
      const { result } = await response.json()
      setAllowances(result)
    }

    fetchData()
  }, [])

  return (
    <div className='p-7'>
      <h2 className='text-green-500 font-semibold text-2xl mb-4'>Allowances</h2>
      <div className='flex flex-wrap gap-3'>
        {allowances.map((allowance) => (
          <Card key={allowance.id} allowance={allowance}/>
        ))}
      </div>
    </div>
  )
}

export default List
