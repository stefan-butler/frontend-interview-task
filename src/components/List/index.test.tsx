import { render, waitFor } from '@testing-library/react'
import List from '@/components/List'
import '@testing-library/jest-dom'
import 'jest-fetch-mock'
import data from '@/data/data.json'
import { Allowance } from '@/types/allowances'

//mock card component 
interface CardProps {
  allowance: Allowance
}

jest.mock('../Card', () => {
  const MockCard = ({ allowance }: CardProps) => {
    return <div data-testid="card">{allowance.name}</div>
  }
  MockCard.displayName = 'MockCard'
  return MockCard
})

describe('List', () => {
  beforeEach(() => {
    fetchMock.mockResponse(JSON.stringify({ result: data }), { status: 200 })
  })

  it('renders as expected', async () => {
    const { container } = render(<List />)
    await waitFor(() => expect(container).toMatchSnapshot())
  })

  it('renders the correct number of cards', async () => {
    const { getAllByTestId } = render(<List />)
    await waitFor(() => {
      const cards = getAllByTestId('card')
      expect(cards).toHaveLength(data.length)
    })

    it('renders allowance details correctly', async () => {
      const { getByText } = render(<List />)
      await waitFor(() => {
        expect(getByText(data[0].name)).toBeInTheDocument()
        expect(getByText(data[1].name)).toBeInTheDocument()
      })
    })
  })
})
