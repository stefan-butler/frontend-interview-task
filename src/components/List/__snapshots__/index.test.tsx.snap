// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List renders as expected 1`] = `
<div>
  <div
    class="p-7"
  >
    <h2
      class="text-green-500 font-semibold text-2xl mb-4"
    >
      Allowances
    </h2>
    <div
      class="flex flex-wrap gap-3"
    />
  </div>
</div>
`;
